File: web/tests/reference/sliding-window/permutation-in-string.js
function checkInclusion(s1, s2) {
  const k = s1.length;
  if (k > s2.length) return false;
  const a = "a".charCodeAt(0);

  const need = new Array(26).fill(0);
  const window = new Array(26).fill(0);
  for (const ch of s1) need[ch.charCodeAt(0) - a]++;

  let matches = need.filter((count) => count === 0).length;

  const bump = (idx, delta) => {
    if (window[idx] === need[idx]) matches--;
    window[idx] += delta;
    if (window[idx] === need[idx]) matches++;
  };

  for (let i = 0; i < k; i++) bump(s2.charCodeAt(i) - a, 1);
  if (matches === 26) return true;

  for (let right = k; right < s2.length; right++) {
    bump(s2.charCodeAt(right - k) - a, -1);
    bump(s2.charCodeAt(right) - a, 1);
    if (matches === 26) return true;
  }

  return false;
}

File: web/tests/reference/sliding-window/permutation-in-string.py
def check_inclusion(s1, s2):
    k = len(s1)
    if k > len(s2):
        return False
    a = ord("a")

    need = [0] * 26
    window = [0] * 26
    for ch in s1:
        need[ord(ch) - a] += 1

    matches = sum(1 for count in need if count == 0)

    def bump(idx, delta):
        nonlocal matches
        if window[idx] == need[idx]:
            matches -= 1
        window[idx] += delta
        if window[idx] == need[idx]:
            matches += 1

    for i in range(k):
        bump(ord(s2[i]) - a, 1)

    if matches == 26:
        return True

    for right in range(k, len(s2)):
        bump(ord(s2[right - k]) - a, -1)
        bump(ord(s2[right]) - a, 1)
        if matches == 26:
            return True

    return False
